### PR TITLE
Aggregate changelogs

### PR DESCRIPTION
**Pull request by saiskee**
# Description

This PR adds links to the appropriate gloo OSS version from the enterprise changelog docs.

# Context

When looking at enterprise changelogs, it is useful for customers and field engineers to have reference to which gloo OSS version the enterprise version relies on. This adds a link to the enterprise changelog docs next to each version.

# Screenshots
image

# Checklist:

- [x] I included a concise, user-facing changelog (for details, see https://github.com/solo-io/go-utils/tree/master/changelogutils) which references the issue that is resolved.
- [x] If I updated APIs (our protos) or helm values, I ran `make install-go-tools generated-code` to ensure there will be no code diff
- [x] I followed guidelines laid out in the Gloo Edge [contribution guide](https://docs.solo.io/gloo-edge/latest/contributing/)
- [x] I opened a draft PR or added the work in progress label if my PR is not ready for review
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/3227

### CODE REVIEW

**Review by kdorosh — CHANGES_REQUESTED** [review]
it appears the open source chronological ordering is now broken

also the enterprise chronologic order is also broken, and doesn't have the aggregate changelog

File: docs/cmd/generate_changelog_doc.go [hunk 1]
```diff
@@ -194,24 +237,293 @@ func getAllReleases(client *github.Client, repo string) ([]*github.RepositoryRel
...
+			releaseA, releaseB := allReleases[i], allReleases[j]
+			versionA, err := versionutils.ParseVersion(releaseA.GetTagName())
+			if err != nil {
+				return false
+			}
+			versionB, err := versionutils.ParseVersion(releaseB.GetTagName())
+			if err != nil {
+				return false
+			}
+			return Version(*versionA).LessThan(Version(*versionB))
```
Comment (kdorosh) [inline docs/cmd/generate_changelog_doc.go:252]: why not,
```suggestion
			return versionB.MustIsGreaterThan(*versionA)
```
and avoid re-declaring types for `Version` and `Versions` at the bottom of the file?

File: docs/cmd/generate_changelog_doc.go [hunk 2]
```diff
@@ -194,24 +237,293 @@ func getAllReleases(client *github.Client, repo string) ([]*github.RepositoryRel
...
+
+	for index, release := range enterpriseReleases {
 		var releaseTag = release.GetTagName()
+
 		version, err := versionutils.ParseVersion(releaseTag)
+		var previousVersion *versionutils.Version
+		// Get the Gloo OSS version that the Gloo enterprise version relied on
+		if index+1 != len(enterpriseReleases) {
+			previousRelease := enterpriseReleases[index+1]
+			previousVersion, err = versionutils.ParseVersion(previousRelease.GetTagName())
```
Comment (kdorosh) [inline docs/cmd/generate_changelog_doc.go:276]: missing error check here

**Review by kdorosh — COMMENTED** [review]
some of this parsing code is getting pretty complex, we should write some unit tests for it

File: docs/cmd/generate_changelog_doc.go [hunk 3]
```diff
@@ -194,24 +237,293 @@ func getAllReleases(client *github.Client, repo string) ([]*github.RepositoryRel
...
+	if err != nil {
+		return nil, err
+	}
+
+	for index, release := range enterpriseReleases {
 		var releaseTag = release.GetTagName()
+
 		version, err := versionutils.ParseVersion(releaseTag)
+		var previousVersion *versionutils.Version
+		// Get the Gloo OSS version that the Gloo enterprise version relied on
```
Comment (kdorosh) [inline docs/cmd/generate_changelog_doc.go:273]:
```suggestion
		// Get the Gloo Edge OSS version that the previous Gloo Edge Enterprise version relied on
```

File: docs/cmd/generate_changelog_doc.go [hunk 4]
```diff
@@ -194,24 +237,293 @@ func getAllReleases(client *github.Client, repo string) ([]*github.RepositoryRel
...
+	openSourceReleases, err := parseGlooReleases(osReleases, false)
+	if err != nil {
+		return nil, err
+	}
+
+	for index, release := range enterpriseReleases {
 		var releaseTag = release.GetTagName()
+
 		version, err := versionutils.ParseVersion(releaseTag)
+		var previousVersion *versionutils.Version
```
Comment (kdorosh) [inline docs/cmd/generate_changelog_doc.go:272]: rename this to `lastGlooOSSDependency`? i want the name to indicate that the version is for open source gloo, the dependency for the last enterprise version we parsed

File: docs/cmd/generate_changelog_doc.go (hunk 1 above)
Comment (saiskee) [reply docs/cmd/generate_changelog_doc.go:252]: The change is valid. However, I'm using the `sort` library to sort versions, and to sort on a type requires that the type have a `LessThan` function.

File: docs/cmd/generate_changelog_doc.go [hunk 5]
```diff
@@ -194,24 +237,293 @@ func getAllReleases(client *github.Client, repo string) ([]*github.RepositoryRel
...
+
+	source := []byte(enterpriseReleaseNotes)
+
+	// Headers : New Features, Fixes, Helm Changes, Dependency Bumps, CVEs
+	headersParsed := make(map[string]int)
+	var eBufEndOfCurrentSection int
+	var endOfCurrentSectionIdx int
+	var offset int
 
+	// Releaes notes are not nested under the headers in the AST, hence we must keep track of the currentHeader
```
Comment (kdorosh) [inline docs/cmd/generate_changelog_doc.go:327]: typo, 
```suggestion
	// Release notes are not nested under the headers in the AST, hence we must keep track of the currentHeader
```

File: docs/cmd/generate_changelog_doc.go [hunk 6]
```diff
@@ -194,24 +237,293 @@ func getAllReleases(client *github.Client, repo string) ([]*github.RepositoryRel
...
+	return minorReleaseMap, nil
+}
+
+func parseEnterpriseNotes(enterpriseReleaseNotes string, openSourceReleases map[Version]string, depVersions []Version) (string, error) {
+	node := goldmark.DefaultParser().Parse(text.NewReader([]byte(enterpriseReleaseNotes)))
+
+	source := []byte(enterpriseReleaseNotes)
+
+	// Headers : New Features, Fixes, Helm Changes, Dependency Bumps, CVEs
+	headersParsed := make(map[string]int)
```
Comment (kdorosh) [inline docs/cmd/generate_changelog_doc.go:322]: we should make this an "enum" for values 1 and 2, so it's more readable. e.g. https://github.com/solo-io/solo-projects/blob/cf2e4e3a9c33b0189adb9a0cfae154e1980c1e77/projects/gloo/pkg/plugins/extauth/process_plugins_test.go#L32-L36

File: docs/cmd/generate_changelog_doc.go [hunk 7]
```diff
@@ -194,24 +237,293 @@ func getAllReleases(client *github.Client, repo string) ([]*github.RepositoryRel
...
+						return "", err
+					}
+					// Get release notes from previous version for current header, and accumulate them
+					if items := osReleaseMap[currentHeader]; len(items) != 0 {
+						for i := 0; i < len(items); i++ {
+							listItem := items[i]
+							n := listItem.FirstChild().Lines().At(0)
+							noteToAppend := osReleaseBuf[n.Start:n.Stop]
+							osReleaseId := strings.ReplaceAll(depVersion.String(), ".", "")
+							osRefLink := fmt.Sprintf("(From [OSS %s](/reference/changelog/open_source/#%s)) ", depVersion.String(), osReleaseId)
```
Comment (kdorosh) [inline docs/cmd/generate_changelog_doc.go:361]: this logic is repeated a couple times, thinking we should pull into a shared function

File: docs/cmd/generate_changelog_doc.go [hunk 8]
```diff
@@ -194,24 +237,293 @@ func getAllReleases(client *github.Client, repo string) ([]*github.RepositoryRel
...
+				return nil, err
+			}
+		}
+
+		minorReleaseMap[minorVersion] = minorReleaseMap[minorVersion] + fmt.Sprintf("##### %s %s\n ", version.String(), glooOSSDescription) + body
 	}
+	return minorReleaseMap, nil
+}
+
+func parseEnterpriseNotes(enterpriseReleaseNotes string, openSourceReleases map[Version]string, depVersions []Version) (string, error) {
```
Comment (kdorosh) [inline docs/cmd/generate_changelog_doc.go:316]: i had a tough time grokking this, would like to see some more comments. perhaps we talk about this function on a call after the other comments are resolved?

Comment (kdorosh) [reply docs/cmd/generate_changelog_doc.go:316]: this is also a case where i think tests would have helped a lot. then i could step through the code easily too.

i think tests are going to be very important with this dense code as i am a bit afraid of brittle code in our changelog generation breaking future docs updates (particularly if the release notes get modified by hand)

File: docs/cmd/changelogutils/changelog_utils.go [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+			if err != nil {
+				return false
+			}
+			return versionA.MustIsGreaterThan(*versionB)
+		})
+	}
+	return allReleases, nil
+}
+
+// Parses OSS Gloo Edge releases into correct format for printing
```
Comment (kdorosh) [inline docs/cmd/changelogutils/changelog_utils.go:48]: we also call this function in the enterprise case, can make the comment more generic

Comment (kdorosh) [reply docs/cmd/changelogutils/changelog_utils.go:48]: in fact, we could rename the function to remove "Gloo" from it's name altogether

**Review by kdorosh — COMMENTED** [review]
in the enterprise case, appears we are missing the version headers (after v1.6) for the other minor releases

File: docs/cmd/test/generate_changelog_doc_test.go [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+			Expect(previousVersion).To(BeNil())
+			previousVersion = changelogutils.GetPreviousEnterprisePatchVersion(enterpriseReleasesSorted, len(sortedVersionsArray)-1)
+			Expect(previousVersion).To(BeNil())
+		})
+
+		It("gets all versions between the current dependency and previous dependency", func() {
+			dependencies := changelogutils.GetAllOSSDependenciesBetweenEnterpriseVersions(&v1_5_5, &v1_5_0_beta8, enterpriseReleasesSorted)
+			expectedDependencies := []Version{v1_5_5, v1_5_0, v1_5_0_beta9}
+			Expect(dependencies).To(Equal(expectedDependencies))
+			//fmt.Printf("%+v\n", dependencies)
```
Comment (kdorosh) [inline docs/cmd/test/generate_changelog_doc_test.go:74]: remove

**Review by kdorosh — CHANGES_REQUESTED** [review]
going to http://localhost:1313/reference/changelog/ and clicking the link to changelogs on the page (rather than the menu on left) redirects to invalid pages with `nil` in them, e.g. http://localhost:1313/reference/changelog/%3cnil%3e/reference/changelog/open_source/

**Review by kdorosh — CHANGES_REQUESTED** [review]
the previous version logic for importing open source changes to enterprise doesn't work in the case of minor version upgrades.

e.g. 1.6.0-beta1 enterprise doesn't have any changes from 1.6.0-beta1 open source. we should fix and add a test for this

File: docs/cmd/changelogutils/changelog_utils.go [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+
+		// Get the Gloo OSS version that the Gloo enterprise version relies on
+		depVersion, err := GetOSSDependencyForEnterpriseVersion(version)
+		var OSSDescription string
+		if err == nil && previousEnterprisePatch != nil {
+			OSSDescription = GetOSDependencyPrefix(*depVersion, true)
+		}
+
+		//TODO: Implement a cache for gloo enterprise dependencies to reduce the amount of API calls
+		previousDepVersion, _ := GetOSSDependencyForEnterpriseVersion(previousEnterprisePatch)
```
Comment (kdorosh) [inline docs/cmd/changelogutils/changelog_utils.go:100]: add comment to explain why we are swallowing errors here?

File: docs/cmd/changelogutils/changelog_utils.go [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+					if emphasis.Level == 2 {
+						// Header block
+						currentHeader = string(typedNode.Text(releaseNotesBuf))
+					}
+
+				} else {
+					v := typedNode.Lines().At(0)
+					note := prefix + fmt.Sprintf("%s\n", v.Value(releaseNotesBuf))
+					if currentHeader != "" {
+						//any extra text e.g. "This release build has failed", only used for enterprise release notes
```
Comment (kdorosh) [inline docs/cmd/changelogutils/changelog_utils.go:172]: this comment belongs in the else block?

File: docs/cmd/changelogutils/changelog_utils.go [hunk 13]
```diff
@@ -0,0 +1,288 @@
...
+		prefix = " (Uses Gloo Edge"
+	}
+	osReleaseURL := strings.ReplaceAll(openSourceVersion.String(), ".", "")
+	osPrefix := fmt.Sprintf("%s [OSS %s](/reference/changelog/open_source/#%s)) ", prefix, openSourceVersion.String(), osReleaseURL)
+	return osPrefix
+}
+
+func GetPreviousEnterprisePatchVersion(enterpriseReleasesSorted []*github.RepositoryRelease, index int) *Version {
+	var previousEnterpriseVersion *Version
+	currentVersion, _ := ParseVersion(enterpriseReleasesSorted[index].GetTagName())
```
Comment (kdorosh) [inline docs/cmd/changelogutils/changelog_utils.go:208]: why are we swallowing errors?

File: docs/cmd/changelogutils/changelog_utils.go [hunk 14]
```diff
@@ -0,0 +1,288 @@
...
+	return previousEnterpriseVersion
+}
+
+// Get the list of open source versions between open source version that the previous enterprise version used and the current enterprise version uses
+func GetAllOSSDependenciesBetweenEnterpriseVersions(startVersion, endVersion *Version, versionsSorted []*github.RepositoryRelease) []Version {
+	var dependencies []Version
+
+	var adding bool
+	for _, release := range versionsSorted {
+		tag, _ := ParseVersion(release.GetTagName())
```
Comment (kdorosh) [inline docs/cmd/changelogutils/changelog_utils.go:230]: why swallow errors?

File: docs/cmd/changelogutils/changelog_utils.go [hunk 15]
```diff
@@ -0,0 +1,288 @@
...
+		return nil, err
+	}
+	return version, nil
+}
+
+// Sorts a slice of versions in descending order by version e.g. v1.6.1, v1.6.0, v1.6.0-beta9
+func SortReleaseVersions(versions []Version) {
+	sort.Slice(versions, func(i, j int) bool {
+		isGreaterThanOrEqualTo := versions[i].MustIsGreaterThanOrEqualTo(versions[j])
+		return isGreaterThanOrEqualTo
```
Comment (kdorosh) [inline docs/cmd/changelogutils/changelog_utils.go:286]: nit, is no easier to read in two lines, let's make one-liner

**Comment by saiskee** [comment]
> the previous version logic for importing open source changes to enterprise doesn't work in the case of minor version upgrades.
> 
> e.g. 1.6.0-beta1 enterprise doesn't have any changes from 1.6.0-beta1 open source. we should fix and add a test for this

https://github.com/solo-io/gloo/blob/d4fea10aee2053e0e8cd2fb7803c608a9bb5fcae/docs/content/reference/changelog/_index.md#L9
Haven't changed any of the logic here. This may just be an effect of how hugo is rendering the links to the child sections. The same issue happens in master on localhost, but the live docs don't have this issue.

**Comment by saiskee** [comment]
> the previous version logic for importing open source changes to enterprise doesn't work in the case of minor version upgrades.
> 
> e.g. 1.6.0-beta1 enterprise doesn't have any changes from 1.6.0-beta1 open source. we should fix and add a test for this

The release build for 1.6.0-beta1 failed and therefore the script can't fetch the OSS dependency for this verison. However, I noticed that "This release build failed" text wasn't rendering, so I fixed the bug and also accounted for that in the tests.

**Review by kdorosh — APPROVED** [review]
nice!

**Pull request merged**